certbot/auth_handler.py

```diff
@@ -356,24 +358,27 @@ def _challenge_factory(self, authzr, path):
         for index in path:
```

Comment:
Rather than modifying this method and the ones it calls, are we able to just modify `_chose_challenges` to not select any challenges for valid authorizations?

There may be reasons to prefer the approach you have here that I haven't thought of, but skipping the logic of choosing which challenges to use and creating all the necessary objects seems like a small win if we can get away with it.

Reply:
My approach was not optimal, and based to a not-optimal comprehension of the logic. Now I got it. Indeed your approach is simpler, as it skips from the source all the logic for challenges that are already validated. I reverted also the log message you commented above, and display it only if there is effectively pending challenges to process.